README.md

[![JHU Data Science](http://jhudatascience.org/images/logo.png)](http://jhudatascience.org/)

[![Travis build status](https://travis-ci.org/muschellij2/fedreporter.svg?branch=master)](https://travis-ci.org/muschellij2/fedreporter) [![AppVeyor Build Status](https://ci.appveyor.com/api/projects/status/github/muschellij2/fedreporter?branch=master&svg=true)](https://ci.appveyor.com/project/muschellij2/fedreporter) [![Coverage status](https://coveralls.io/repos/github/muschellij2/fedreporter/badge.svg?branch=master)](https://coveralls.io/r/muschellij2/fedreporter?branch=master) <!-- README.md is generated from README.Rmd. Please edit that file -->

fedreporter Package:
====================

The goal of `fedreporter` is to provide downloads data from NIH 'ExPORTER'

Installation
------------

You can install `fedreporter` from GitHub with:

``` r
# install.packages("remotes")
remotes::install_github("muschellij2/fedreporter")
```

Example
=======

Project Search for R01s from NIH
--------------------------------

``` r
library(fedreporter)
res = fe_projects_search(
  project_number = "*R01*",
  fiscal_year = 2012,
  agency = "NIH",
  text = "stroke",
  text_field = "title")
#> GET command is:
#> Response [https://api.federalreporter.nih.gov/v1/Projects/search?query=projectNumber%3A*R01*%24fy%3A2012%24agency%3ANIH%24text%3Astroke%24textFields%3Atitle%24textOperator%3AAND&offset=1&limit=50]
#>   Date: 2017-10-03 15:33
#>   Status: 200
#>   Content-Type: application/json; charset=utf-8
#>   Size: 249 kB
names(res)
#> [1] "response" "content"
res$response
#> Response [https://api.federalreporter.nih.gov/v1/Projects/search?query=projectNumber%3A*R01*%24fy%3A2012%24agency%3ANIH%24text%3Astroke%24textFields%3Atitle%24textOperator%3AAND&offset=1&limit=50]
#>   Date: 2017-10-03 15:33
#>   Status: 200
#>   Content-Type: application/json; charset=utf-8
#>   Size: 249 kB
length(res$content)
#> [1] 4
names(res$content)
#> [1] "totalCount" "offset"     "limit"      "items"
res$content$offset
#> [1] 1
length(res$content$items)
#> [1] 50
```

Project Search for Individual PIs
---------------------------------

``` r
res = fe_projects_search(pi_name = "MATSUI, ELIZABETH")
#> GET command is:
#> Response [https://api.federalreporter.nih.gov/v1/Projects/search?query=piName%3AMATSUI%2C%20ELIZABETH&offset=1&limit=50]
#>   Date: 2017-10-03 15:33
#>   Status: 200
#>   Content-Type: application/json; charset=utf-8
#>   Size: 128 kB
items = res$content$items
con_pis = sapply(items, "[[", "contactPi")
keep = grepl("^MATSUI", con_pis)
items = items[keep]
mat_costs = sapply(items, "[[", "totalCostAmount")
sum(mat_costs)
#> [1] 16021336
                         
res = fe_projects_search(pi_name = "PENG, ROGER")
#> GET command is:
#> Response [https://api.federalreporter.nih.gov/v1/Projects/search?query=piName%3APENG%2C%20ROGER&offset=1&limit=50]
#>   Date: 2017-10-03 15:33
#>   Status: 200
#>   Content-Type: application/json; charset=utf-8
#>   Size: 55 kB
items = res$content$items
con_pis = sapply(items, "[[", "contactPi")
keep = grepl("^PENG", con_pis)
items = items[keep]
peng_costs = sapply(items, "[[", "totalCostAmount")
sum(peng_costs)
#> [1] 2868853
                         
# both = fe_projects_search(pi_name = c("MATSUI, ELIZABETH", "PENG, ROGER"))
```


NEWS.md
# fedreporter 0.2.1

* Resubmission to CRAN after typos.

# fedreporter 0.2.0

* Added a `NEWS.md` file to track changes to the package.

* First submission. 





tests/testthat.R
library(testthat)
library(fedreporter)

test_check("fedreporter")


tests/testthat/test-projects.R
test_that("Check projects", {
  L = fe_projects(project_number = "5R01MH092950-05")
  expect_equal(httr::status_code(L$response), 200)

  expect_error(fe_projects(), regexp = "must specify")
  expect_error(fe_projects_multiple(), regexp = "must be")
})



R/fe_projects.R

#' Federal RePORTER projects
#'
#' @param nih_id global identifier for a project across all NIH
#' systems that handle research projects/grants data.
#' Example, 8828294.
#' @param project_number unique number that is assigned to a project
#'  by the affiliated federal agency. Examples: \code{'5R01MH092950-05'},
#'  \code{'1R01CA183929-01A1'}, \code{'USFS-0000779'}.
#' @param internal_id internal Federal RePORTER generated unique ID
#' associated with each project. It is available as part of
#' export on Federal RePORTER search results. Example, 739576.
#' @param secure passed to \code{\link{fe_base_url}} for https
#' @note See \url{https://api.federalreporter.nih.gov/#!/Projects/Get_Project}
#' @param verbose print diagnostic messages
#'
#' @return List of the result of the \code{\link{GET}} call and
#' the content
#' @export
#'
#' @importFrom httr GET
#' @examples
#' res = fe_projects(project_number = "5R01MH092950-05")
fe_projects = function(
  nih_id = NULL,
  internal_id = NULL,
  project_number = NULL,
  verbose = TRUE,
  secure = TRUE) {

  url = fe_base_url(secure = secure)
  path = "/v1/Projects"
  url = paste0(url, path)

  query = list()
  query$smApplId = internal_id
  query$nihApplId = nih_id
  query$projectNumber = project_number

  if (length(query) == 0) {
    query = NULL
    stop("No projects have been specified - must specify one!")
  }
  query = lapply(query, as.character)

  res = httr::GET(url, query = query)
  if (verbose) {
    message("GET command is:")
    print(res)
  }
  httr::stop_for_status(res)
  cr = httr::content(res)

  L = list(
    response = res,
    content = cr)
  return(L)
}
#
# smApplID: It is an
# nihApplId: It is a
# projectNumber: It is a


R/fe_projects_multiple.R
#' Fetch multiple projects by internal Federal RePORTER IDs (SM ID
#' or Project Numbers)
#'
#' @param internal_id internal Federal 'RePORTER' generated unique ID
#' associated with each project. It is available as part of
#' export on Federal RePORTER search results. Example, 739576.
#' @param project_number unique number that is assigned to a project
#'  by the affiliated federal agency. Examples: \code{'5R01MH092950-05'},
#'  \code{'1R01CA183929-01A1'}, \code{'USFS-0000779'}.
#' @param secure passed to \code{\link{fe_base_url}} for https
#' @note See
#' \url{https://api.federalreporter.nih.gov/#!/Projects/Get_Project},
#' this calls \code{POST /v1/projects/FetchBySmApplIds}
#'
#' @return List of the result of the \code{\link{GET}} call and
#' the content
#' @export
#' @importFrom httr POST stop_for_status content content_type_json
#' @importFrom jsonlite toJSON
#' @examples
#' res = fe_projects_multiple(
#' project_number = c("5R01MH092950-05", "USFS-0000779")
#' )
#' res = fe_projects_multiple(
#' internal_id = c("739576", "739577")
#' )
fe_projects_multiple = function(
  internal_id = NULL,
  project_number = NULL,
  secure = TRUE) {

  url = fe_base_url(secure = secure)
  if (is.null(internal_id) && is.null(project_number)) {
    stop("Internal IDs or Project Numbers must be specified!")
  }
  if (!is.null(internal_id) && is.null(project_number)) {
    fetch_by = "SmApplIds"
    body = internal_id
  }
  if (is.null(internal_id) && !is.null(project_number)) {
    fetch_by = "ProjectNumbers"
    body = project_number
  }

  path = paste0("/v1/Projects/", "FetchBy", fetch_by)
  url = paste0(url, path)

  # if (length(query) == 0) {
  #   query = NULL
  #   stop("No projects have been specified - must specify one!")
  # }
  # query = lapply(query, function(x) {
    # jsonlite::toJSON(x)
  # })
  body = jsonlite::toJSON(body)

  res = httr::POST(url, body = body, httr::content_type_json())
  httr::stop_for_status(res)
  cr = httr::content(res)

  L = list(
    response = res,
    content = cr)
  return(L)
}


R/fe_projects_search.R
#' Search Federal RePORTER Projects
#'
#' @param project_number unique number that is assigned to a project
#'  by the affiliated federal agency. Examples, 5R01MH092950-05,
#'  1R01CA183929-01A1, USFS-0000779.
#'  Can also use wildcards such as \code{*R01*}.
#' @param fiscal_year Fiscal year of data to obtain
#' @param text Search text from title, abstract, and terms data
#' @param text_field text field to search, can be title, abstract or terms,
#' and defaults to search all. Only enabled when \code{text} is not \code{NULL}.
#' @param text_operator Text operator to combine text terms,
#' Can be AND or OR, AND is default.
#' Only enabled when \code{text} is not \code{NULL}.
#' @param agency Agency code to search.  see \code{\link{fe_agencies}}
#' @param pi_name Principal investigator name.  Names are to be in the
#' \code{Last First M}, \code{Last, First}, \code{First},
#' \code{Last}, or \code{First Last} formats.  Multiple names can be
#' given.
#' @param offset start at item # (starts at 1)
#' @param limit max # of items to return (at the most 50 per request)
#' @param secure Should https be used, passed to \code{\link{fe_base_url}}
#' @param verbose print diagnostic messages
#'
#' @return List of the result of the \code{\link{GET}} call and
#' the content
#' @export
#'
#' @examples
#' res = fe_projects_search(
#' project_number = "USFS*",
#' fiscal_year = 2012)
#' res = fe_projects_search(
#' project_number = "*R01*",
#' fiscal_year = 2012,
#' agency = "NIH",
#' text = "stroke",
#' text_field = "title")
#'
#' \dontrun{
#' res = fe_projects_search(pi_name = "MATSUI, ELIZABETH")
#' items = res$content$items
#' con_pis = sapply(items, "[[", "contactPi")
#' keep = grepl("^MATSUI", con_pis)
#' items = items[keep]
#' mat_costs = sapply(items, "[[", "totalCostAmount")
#' sum(mat_costs)
#'
#' res = fe_projects_search(pi_name = "PENG, ROGER")
#' items = res$content$items
#' con_pis = sapply(items, "[[", "contactPi")
#' keep = grepl("^PENG", con_pis)
#' items = items[keep]
#' peng_costs = sapply(items, "[[", "totalCostAmount")
#' sum(peng_costs)
#'
#' both = fe_projects_search(
#' pi_name = c("MATSUI, ELIZABETH", "PENG, ROGER"))
#' }
fe_projects_search = function(
  project_number = NULL,
  fiscal_year = NULL,
  text = NULL,
  text_field = c("title", "abstract", "terms"),
  text_operator = c("AND", "OR"),
  agency = NULL,
  pi_name = NULL,
  offset = 1,
  limit = 50,
  verbose = TRUE,
  secure = TRUE
){

  url = fe_base_url(secure = secure)
  path = "/v1/Projects/search"
  url = paste0(url, path)

  query = list()


  # projectNumber
  if (!is.null(project_number)) {
    project_number = paste(project_number, collapse = ",")
    query$projectNumber = project_number
  }

  # fy
  if (!is.null(fiscal_year)) {
    fiscal_year = as.integer(fiscal_year)
    fiscal_year = paste(fiscal_year, collapse = ",")
    query$fy = fiscal_year
  }

  # agency
  if (!is.null(agency)) {
    fields = fedreporter::fe_agencies
    agency = match.arg(agency, choices = fields, several.ok = TRUE)
    agency = paste(agency, collapse = ",")
    query$agency = agency
  }

  # text
  if (!is.null(text)) {
    text = paste(text, collapse = " ")
    query$text = text

    # textFields
    if (!is.null(text_field)) {
      fields = c("title", "abstract", "terms")
      text_field = match.arg(text_field, choices = fields, several.ok = TRUE)
      text_field = paste(text_field, collapse = ",")
      query$textFields = text_field
    }

    # textOperator:AND
    if (!is.null(text_operator)) {
      text_operator = match.arg(text_operator)
      query$textOperator = text_operator
    }
  }

  # piName
  if (!is.null(pi_name)) {
    pi_name = paste(pi_name, collapse = ";")
    pi_name = toupper(pi_name)
    query$piName = pi_name
  }

  if (length(query) == 0) {
    stop("query has no search terms!")
  }
  nquery = names(query)
  query = mapply(function(x, n) {
    paste0(n, ":", x)
  }, query, nquery, SIMPLIFY = TRUE)
  query = paste(query, collapse = "$")
  query = gsub(":", "%3A", query)
  query = gsub("$", "%24", query, fixed = TRUE)
  query = gsub(" ", "%20", query)
  query = gsub(",", "%2C", query)


  query = list(
    query = I(query)
    )
  query$offset = offset
  query$limit = limit

  # new_url <- httr::parse_url(url)
  # new_url$query = query
  # new_url$query$query =  I(new_url$query$query)
  # new_url = httr::build_url(new_url)
  # new_url = gsub("%2A", "*", new_url)
  # new_url = gsub("%2C", ",", new_url)
  # new_url = gsub("%24", "$", new_url)
  # new_url = gsub(":", "%3A", new_url)
  # print(new_url)
  # res = httr::GET(new_url, httr::accept_json())


  res = httr::GET(url, query = query)
  if (verbose) {
    message("GET command is:")
    print(res)
  }
  httr::stop_for_status(res)
  cr = httr::content(res)

  L = list(
    response = res,
    content = cr)
  return(L)

}
